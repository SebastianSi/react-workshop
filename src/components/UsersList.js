import React, { Component } from 'react';
import UserListItem from './UserListItem';
import UserForm from './UserForm';
import mockApi from '../utils/mockApi';
import { withStyles } from '@material-ui/core/styles';
import Button from './common/Button';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { usersAction } from '../actions/usersActions';

class UsersList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      users: [],
      isFormOpen: false,
      isAddMode: false,
      currentUserClicked: null
    };
  }

  openUserForm = (userId, isAddMode) => {
    this.setState({
      currentUserClicked: userId,
      isFormOpen: true,
      isAddMode
    });
  };

  closeUserForm = () => {
    this.setState({
      currentUserClicked: null,
      isFormOpen: false
    });
  };

  onUserFormSubmit = user => {
    console.log(user);
    if (this.state.isAddMode) {
      mockApi.addUser(user).then(this.fetchUsers);
    } else {
      mockApi.updateUser(user).then(this.fetchUsers);
    }
    this.closeUserForm();
  };

  onUserDelete = userId => {
    mockApi.deleteUser(userId).then(this.fetchUsers);
    this.closeUserForm();
  };

  componentDidMount() {
    this.fetchUsers();
  }

  fetchUsers = () => {
    // just call 'this.props.getUsers()' without any api call, remove the fetch call
    mockApi.fetchUsers().then(users => {
      this.props.getUsers(users);
      // this.setState({ users });
    });
  };

  renderUsersList = users => (
    <ul className={'users-list-container'}>
      {users.map(user => (
        <UserListItem key={user.id} user={user} openUserForm={this.openUserForm} />
      ))}
    </ul>
  );

  render() {
    const { classes, users } = this.props;
    let { isAddMode, isFormOpen, currentUserClicked } = this.state;
    console.log('users--->', users);
    return (
      // <div className={classes.root}>
      <div>
        {isFormOpen ? (
          <UserForm
            isAddMode={isAddMode}
            userId={currentUserClicked}
            onSubmit={this.onUserFormSubmit}
            onDelete={this.onUserDelete}
            onCancel={this.closeUserForm}
          />
        ) : (
          <>
            {!!users.length && this.renderUsersList(users)}
            <div>
              <Button
                text={'Add user'}
                onClick={() => {
                  this.openUserForm(null, true);
                }}
              />
              <Link to={'/contact'}>
                {' '}
                <Button text={'Contact Page'} />{' '}
              </Link>
              <Link to={'/posts'}>
                {' '}
                <Button text={'Posts Page'} />{' '}
              </Link>
            </div>
          </>
        )}
      </div>
    );
  }
}

const styles = () => ({
  root: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    width: 450,
    height: 700,
    borderRadius: 5
  }
});

const mapStateToProps = state => ({
  users: state.usersReducer.users
});

const mapDispatchToProps = dispatch => ({
  // remove users param
  getUsers: users => {
    dispatch(usersAction(users));
  }
});

// export default withStyles(styles)(UsersList);
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(UsersList);
