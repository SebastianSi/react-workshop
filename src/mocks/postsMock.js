export const postsMock = [
  {
    title: "title 1",
    description: "description 1",
    imageIndex: "1",
    likes: "4",
  },
  {
    title: "title 2",
    description: "description 2",
    imageIndex: "2",
    likes: "3",
  },
  {
    title: "title 3",
    description: "description 3",
    imageIndex: "2",
    likes: "1",
  },
  {
    title: "title 4",
    description: "description 4",
    imageIndex: "1",
    likes: "4",
  }
]

export const postMock = {
  title: "title",
  description: "description",
  imageIndex: "4",
  likes: "5",
}