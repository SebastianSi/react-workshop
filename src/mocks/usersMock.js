export const postsMock = [
  {
    id: 0,
    age: 27,
    email: "ionelcelchel@hotmail.com",
    name: "Ionel"
  },
  {
    id: 1,
    age: 22,
    email: "andrei@hotmail.com",
    name: "Andrei"
  },
  {
    id: 2,
    age: 17,
    email: "testuser@hotmail.com",
    name: "Test"
  },
  {
    id: 3,
    age: 31,
    email: "user@hotmail.com",
    name: "User"
  },
];