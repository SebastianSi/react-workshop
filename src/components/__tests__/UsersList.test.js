import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { BrowserRouter as Router } from 'react-router-dom';

import { UsersList } from '../UsersList';
import { usersMock } from '../../mocks/usersMock';

test('UsersList snapshot', () => {
  const getUsersMockFn = jest.fn();
  const component = renderer.create(
    <Router>
      <UsersList users={usersMock} getUsers={getUsersMockFn} />
    </Router>
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('UsersList tests', () => {
  const getUsersMockFn = jest.fn();
  const component = shallow(
    <UsersList users={usersMock} getUsers={getUsersMockFn} />
  );

  //called on componentDidMount
  expect(getUsersMockFn).toHaveBeenCalledTimes(1);
});