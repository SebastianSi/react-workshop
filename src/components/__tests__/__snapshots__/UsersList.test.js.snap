// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersList snapshot 1`] = `
<div>
  <ul
    className="users-list-container"
  >
    <li
      style={
        Object {
          "padding": 10,
        }
      }
    >
      <span
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        name: 
        <b>
          Ionel
        </b>
         
      </span>
      <button
        onClick={[Function]}
      >
        Edit
      </button>
    </li>
    <li
      style={
        Object {
          "padding": 10,
        }
      }
    >
      <span
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        name: 
        <b>
          Andrei
        </b>
         
      </span>
      <button
        onClick={[Function]}
      >
        Edit
      </button>
    </li>
    <li
      style={
        Object {
          "padding": 10,
        }
      }
    >
      <span
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        name: 
        <b>
          Test
        </b>
         
      </span>
      <button
        onClick={[Function]}
      >
        Edit
      </button>
    </li>
    <li
      style={
        Object {
          "padding": 10,
        }
      }
    >
      <span
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        name: 
        <b>
          User
        </b>
         
      </span>
      <button
        onClick={[Function]}
      >
        Edit
      </button>
    </li>
  </ul>
  <div>
    <button
      onClick={[Function]}
    >
      Add user
    </button>
    <a
      href="/contact"
      onClick={[Function]}
    >
       
      <button>
        Contact Page
      </button>
       
    </a>
    <a
      href="/posts"
      onClick={[Function]}
    >
       
      <button>
        Posts Page
      </button>
       
    </a>
  </div>
</div>
`;
